### ISSUE TITLE
feat: allow the index as an argument to list functions such as filter and map

### ISSUE THREAD

**Issue by stephen-bowser** [ibis-project/ibis#10201]
### Is your feature request related to a problem?

Not a problem so much as an enhancement.

### What is the motivation behind your request?

Backends like pyspark and duckdb allow you to use the index of the element as an argument in the function calls to array functions like map and filter. (e.g.: for [duckdb](https://duckdb.org/docs/sql/functions/lambda.html#indexes-as-parameters), [pyspark](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.transform.html)). 

Having the index of the element is helpful when e.g.: you want to filter a list to take every 3rd element.

### Describe the solution you'd like

Allow an optional second argument to functions like [map](https://ibis-project.org/reference/expression-collections#ibis.expr.types.arrays.ArrayValue.map) and [filter](https://ibis-project.org/reference/expression-collections#ibis.expr.types.arrays.ArrayValue.filter) which would reflect the index of the element.

I can make workarounds for this, by creating an additional column containing the indexes, then running an array_zip between that and my original list to create a struct but it's not as clean

### What version of ibis are you running?

9.2.0

### What backend(s) are you using, if any?

DuckDB, PySpark

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by cpcloud** [comment]
Thanks for the issue!

Seems like a reasonable ask. I'll put up a PR shortly, for discussion.

**Comment by stephen-bowser** [comment]
Thank you for resolving this so quickly! I've tested it on one of your dev releases (10.0.0.dev148) and it seems to work well :)
Can I check what your release schedule looks like and when this might make it into a production release?

**Comment by cpcloud** [comment]
@stephen-bowser Hard to say exactly, but hopefully by November.

**Comment by stephen-bowser** [comment]
> @stephen-bowser Hard to say exactly, but hopefully by November we'll have released 10.0.0.

Thank you!

As an update, I did some more testing with this and spotted some unexpected behaviour when using this in pyspark.

With duckdb, which works as expected:

```
import duckdb
import ibis

DUCKDB_CONNECTION_NAME = "test.duckdb"

con = duckdb.connect(DUCKDB_CONNECTION_NAME)
con.execute("CREATE OR REPLACE TABLE test_array AS SELECT [1, 2, 3, null, null, null, null] AS arr")


ibis_duckdb_con = ibis.duckdb.connect(DUCKDB_CONNECTION_NAME)

ibis_duckdb_table = ibis_duckdb_con.table('test_array')
ibis_duckdb_table.arr.filter(lambda x, i: i % 4 == 0)
# -> returns [1, None]
```

Whereas with pyspark, the filter seems to remove null values.

```
from pyspark.sql import SparkSession
spark = SparkSession.builder.getOrCreate()

spark.sql("CREATE OR REPLACE TEMPORARY VIEW test_array AS SELECT array(1, 2, 3, null, null, null, null) AS arr")

ibis_pyspark_con = ibis.pyspark.connect(spark)

ibis_pyspark_table = ibis_pyspark_con.table('test_array')

pyspark_output = ibis_pyspark_table.mutate(array_output = ibis_pyspark_table.arr.filter(lambda x, i: i % 4 == 0))

spark.sql(pyspark_output.compile()).select('array_output').show()

# -> returns [1]
```

Looking at the compiled sql, it looks like this issue is that the pyspark version is first doing a `transform` to map records which don't meet the filter condition to nulls, and then filtering out null records. However, this means that any records which were already null get caught in the crossfire.

compiled pyspark sql
```
select
    `t0`.`arr`, filter (
        transform(
            `t0`.`arr`,
            (__ibis_param_x__, __ibis_param_i__)
            -> if((__ibis_param_i__ % 4) = 0, __ibis_param_x__, null)
        ),
        (__ibis_param_x__, __ibis_param_i__) -> __ibis_param_x__ is not null
    ) as `array_output`
from `test_array` as `t0`
```

For comparision, the duckdb version just directly filters based on the condition.
```
select
    "t0"."arr",
    list_filter(
        "t0"."arr",
        (__ibis_param_x__, __ibis_param_i__) -> ((__ibis_param_i__ - 1) % 4) = 0
    ) as "array_output"
from "test_array" as "t0"
```

**Comment by stephen-bowser** [comment]
Just bumping this in case it got missed.
